lint/mypy.bzl

```diff
@@ -0,0 +1,190 @@
+
+    See https://mypy.readthedocs.io/en/stable/command_line.html
+
+    Args:
+        ctx: Bazel Rule or Aspect evaluation context
+        executable: label of the the mypy program
+        srcs: python files to be linted
+        deps: the deps of the py_library or py_binary so mypy can read dependent types
+        configs: label(s) of mypy config file(s)
+
+    Returns:
+        Providers, including a MypyInfo provider to propagate data to dependents and a Validation output group
+    """
+    inputs = depset(srcs + configs, transitive = _extract_transitive_inputs(deps))
+    cache_map_outputs = []
+    args = ctx.actions.args()
+    args.add_all(srcs)
+
+    # TODO: do we need to support bazel's "implicit __init__.py convention"?
+    # It seems that `--incompatible_default_to_explicit_init_py` is now preferred based on
+    # https://github.com/bazelbuild/bazel/issues/7386#issuecomment-1806973756
+    #
+    # --package-root adds a directory below which directories are considered
+    # packages even without __init__.py.  May be repeated.
+    # args.add("--package-root")
+
+    # Skip cache internal consistency checks based on mtime
+    args.add("--skip-cache-mtime-checks")
+
+    # Undocumented flag, see https://github.com/python/mypy/pull/4759
+    # The --bazel flag make cache files hermetic by using relative paths and setting mtime to zero;
+    # effectively the presence of a cache file prevents reading of the source file
+    # (though the source file must still exist).
+    args.add("--bazel")
+
+    # Undocumented flag, see https://github.com/python/mypy/pull/4759
+    # The --cache-map flag specifies a mapping from source files to cache files that overrides the
+    # usual way of finding the locations for cache files; e.g.
+    # mypy --cache-map foo.py foo.meta.json foo.data.json -- foo.py
+    # There must be any number of triples (source, meta, data) with the additional constraints that
+    # the source must end in .py or .pyi, meta must end in .meta.json,
+    # and data must end in .data.json. For files in the cache map, the default cache directory
+    # (--cache-dir) is ignored, and if the cache files named in the --cache-map don't exist
+    # they will be written. For files not in the cache map, the --cache-dir is still used.
+    cache_map = []  # Items for cache_map file.
+    for src in srcs:
+        cache_map.append(src)
+
+        # NB: handle //path/to/package:subdir/file.py -> subdir/file.py
+        path = paths.relativize(src.path, ctx.label.package)
+        for kind in ("meta", "data"):
+            # Note, this assumes every source file has a single mypy aspect producing these output files
+            # For subdir/file.py, declare subdir/file.meta.json
+            file = ctx.actions.declare_file(paths.replace_extension(path, ".%s.json" % kind))
+            cache_map_outputs.append(file)
+            cache_map.append(file)
+    args.add_all(["--cache-map"] + cache_map)
```

Comment:
nit: doesn't this copy the list? could do `args.add('--cache-map'); args.add_all(cache_map)` instead